### PR TITLE
Models

### LINKED ISSUE #173: Abstract model cleanup
- [x] Renamed `redirectPath` to something more descriptive
- [x] Cleanup `toJSON` implementations to remove non-editable fields
- [x] Change `notes` datatype from `string` to `object/hash`
  - `notes` inputs are json blobs
  - Potentially need to use this: [class](https://github.com/eligrey/Blob.js) (recommended by [angular](https://angular.io/guide/browser-support#suggested-polyfills))

### PR DESCRIPTION
**Pull request by Allcharles**
# BAW Server Models

Created the basic framework for the BAW server models.

## Changes

Created the following models:

- `AnalysisJob`
- `AnalysisJobItem`
- `AudioEvent`
- `AudioEventComment`
- `AudioEventTag`
- `Bookmark`
- `Dataset`
- `DatasetItem`
- `ProgressEvent`
- `SavedSearches` <- Will rename to SavedSearch in another PR
- `Study`

Renamed `redirectionPath()` to `viewUrl` and `getViewUrl()`

## Issues

Changing models to try use Blobs likely caused some visual bugs on model forms. However this only affects pages which are connected to mock services so it has no affect on functionality.

## Closes 

Closes #173

### CODE REVIEW

**Comment by Allcharles** [comment]
~~Need to change `navigationPath()` to `get viewUrl()`.~~ Not possible as navigationPath sometimes requires inputs.

File: src/app/models/AnalysisJob.ts
Comment (atruskie) [inline src/app/models/AnalysisJob.ts:99]: this should be JSON in DB, and part of object graph in response

File: src/app/models/AnalysisJob.ts
Comment (atruskie) [inline src/app/models/AnalysisJob.ts:98]: This will usually be something like a YAML file - Blob is correct type

File: src/app/models/AnalysisJob.ts
Comment (atruskie) [inline src/app/models/AnalysisJob.ts]: For `create` customSettings is writeable and should be serialized

File: src/app/models/AnalysisJob.ts
Comment (atruskie) [inline src/app/models/AnalysisJob.ts:103]: i'd rather see this error so we catch it early, especially since intend to add this view

File: src/app/models/AnalysisJob.ts
Comment (atruskie) [inline src/app/models/AnalysisJob.ts:108]: Relevant reading: https://github.com/QutEcoacoustics/baw-server/blob/d165c88677ac97f1fa6be47d28b7774b1bed7846/app/models/analysis_job.rb#L118-L149

File: src/app/models/AnalysisJobItem.ts
Comment (atruskie) [inline src/app/models/AnalysisJobItem.ts]: this model should always be read-only

File: src/app/models/AnalysisJobItem.ts
Comment (atruskie) [inline src/app/models/AnalysisJobItem.ts:47]: should throw

File: src/app/models/AnalysisJobItem.ts
Comment (atruskie) [inline src/app/models/AnalysisJobItem.ts:52]: https://github.com/QutEcoacoustics/baw-server/blob/d165c88677ac97f1fa6be47d28b7774b1bed7846/app/models/analysis_jobs_item.rb#L190-L217

File: src/app/models/AudioEvent.ts
Comment (atruskie) [inline src/app/models/AudioEvent.ts:68]: throw

File: src/app/models/AudioEventComment.ts
Comment (atruskie) [inline src/app/models/AudioEventComment.ts]: Remove entire model, deprecated

File: src/app/models/AudioEventTag.ts
Comment (atruskie) [inline src/app/models/AudioEventTag.ts:53]: leave a note: this will probably never have it's own view. Should return `null`?

File: src/app/models/AudioRecording.ts
Comment (atruskie) [inline src/app/models/AudioRecording.ts]: AudioRecordings will probably always be immutable for the client app.

File: src/app/models/AudioRecording.ts
Comment (atruskie) [inline src/app/models/AudioRecording.ts:100]: throw.

File: src/app/models/AudioRecording.ts
Comment (atruskie) [inline src/app/models/AudioRecording.ts:104]: https://github.com/QutEcoacoustics/baw-server/blob/d165c88677ac97f1fa6be47d28b7774b1bed7846/app/models/audio_recording.rb#L39

File: src/app/models/Bookmark.ts
Comment (atruskie) [inline src/app/models/Bookmark.ts:66]: we'll need a `listenViewUrl` function at some point. It'll accept recording/recordingId, and optional start offset

File: src/app/models/Dataset.ts
Comment (atruskie) [inline src/app/models/Dataset.ts:53]: `null` for now - no view planned

File: src/app/models/DatasetItem.ts
Comment (atruskie) [inline src/app/models/DatasetItem.ts:51]: null for now - no view planned

File: src/app/models/ProgressEvent.ts
Comment (atruskie) [inline src/app/models/ProgressEvent.ts:42]: null for now - no view planned

File: src/app/models/SavedSearches.ts
Comment (atruskie) [inline src/app/models/SavedSearches.ts]: throw

File: src/app/models/SavedSearches.ts
Comment (atruskie) [inline src/app/models/SavedSearches.ts]: This will be an object of type `Filter`

File: src/app/models/Study.ts
Comment (atruskie) [inline src/app/models/Study.ts:49]: throw

File: src/app/models/Tag.ts
Comment (atruskie) [inline src/app/models/Tag.ts:46]: https://github.com/QutEcoacoustics/baw-server/blob/d165c88677ac97f1fa6be47d28b7774b1bed7846/app/models/tag.rb#L20-L26

File: src/app/models/Tag.ts
Comment (atruskie) [inline src/app/models/Tag.ts:42]: this field should not be returned by the API?

File: src/app/models/User.ts
Comment (atruskie) [inline src/app/models/User.ts:97]: filed an issue?

File: src/app/models/User.ts
Comment (atruskie) [inline src/app/models/User.ts:197]: not writeable

**Review by atruskie — CHANGES_REQUESTED** [review]
It's great that all the intermediate objects used for view rendering were removed. Removes a lot of fluff and reduces chance for weird-state partial models causing unexpected behaviour.

File: src/app/models/User.ts
Comment (Allcharles) [reply src/app/models/User.ts:97]: I can file one, but changing this behaviour on the baw-server would require making the opposite adjustment on the baw-client (ie. removing /assets from imageUrls). The change will have to wait until both are merged.

**Comment by Allcharles** [comment]
A lot of your suggestions (especially those affecting models) will have to be added in an upstream PR. Unfortunately because this PR is almost 2 weeks old a few of the models you have mentioned have been changed when i implemented decorators. I'll try collate a list of commits relevant to this review below.

File: src/app/models/User.ts
Comment (Allcharles) [reply src/app/models/User.ts:197]: This is required currently because of the `handleAuth` function in the security service. When a user logs in, i get a User model which i then convert into a Session User model: https://github.com/QutEcoacoustics/workbench-client/blob/master/src/app/services/baw-api/security.service.ts#L84-L90. Potentially there is a better was of doing this, however my assumption was that the server would be able to take only the values it needed.

Comment (atruskie) [reply src/app/models/User.ts:197]: The only model that should have the auth token is the current user model. Even then,l it doesn't need it - the auth token shouldn't be part of any User model.

Thus:
- Ensure the current session/user is passed to the security service (injected, not as part of anyhting that will get toJSON'ed)
- access the the `user.authToken` as you would normally.

Comment (Allcharles) [reply src/app/models/User.ts:197]: - The `authToken` is not part of the `User` model. It is only part of the `SessionUser` model.
- I'm not sure what you mean by injecting the current `SessionUser` model into the security service, as the service is what it generating the model
- Removing the functionality from the `toJSON()` in the `User` model would still not work because I need to store the model inside localstorage and `JSON.stringify` will use the output of `toJSON()`

Comment (atruskie) [reply src/app/models/User.ts:197]: Ah local storage serialization. Fair enough, carry on

**Comment by Allcharles** [comment]
Fix AnalysisJob review comments: 93f3e69e7da50be6c70787a50efc6e7da31638a6

File: src/app/models/AudioEventTag.ts
Comment (Allcharles) [reply src/app/models/AudioEventTag.ts:53]: Ill throw an error instead of `null`, it'll be caught easier if a view is ever made.

File: src/app/models/Tag.ts
Comment (Allcharles) [reply src/app/models/Tag.ts:46]: We agreed previously that the type of tags may be changed in the future and may become an api request. This is because of the `typeofTag`s is used as a dropdown field in a couple places.

File: src/app/models/Tag.ts
Comment (Allcharles) [reply src/app/models/Tag.ts:42]: Its shown in the tags list view and i didnt want to create 25 api requests when i display the list of tags. If its not needed, we can remove it from the view: https://staging.ecosounds.org/admin/tags which also removes the issue

Comment (atruskie) [reply src/app/models/Tag.ts:42]: leave it commented - it's not a field we'd unusually return since it is an aggregate value

Comment (Allcharles) [reply src/app/models/Tag.ts:42]: As in comment out the key, or leave a comment explaining that its in limbo? The value is already in use on the tag list page so commenting it out would require me to modify that too.

Comment (atruskie) [reply src/app/models/Tag.ts:42]: Comment in limbo is fine.

This will have to be revisited because I do expect that field will  be output by the API. It contravenes the API conventions.

It's a little different for the rails view: it can do arbitrary queries anywhere anytime - which is what it does. _count_ is not part of the model server side either.

**Review by atruskie — APPROVED** [review]

**Pull request merged**
